File: frontend/src/api/actions/notes.ts
import { fetchJSON, postAction } from '../client';
import type { NoteContent } from '../types';

// GET, not postAction - server.py's /api/note?rel= is a pass-through
// read, no state change (server.py's read_note()). The snapshot's own
// `notes` collector only ever carries previews; this is how the reader/
// editor gets a note's actual body, fetched on demand.
export function fetchNote(rel: string): Promise<NoteContent> {
  return fetchJSON(`/api/note?rel=${encodeURIComponent(rel)}`);
}

export function saveNote(rel: string, text: string): Promise<{ ok: boolean; rel?: string; when?: number }> {
  return postAction('/api/note/save', { rel, text });
}

export function newNote(name: string, folder = ''): Promise<{ ok: boolean; rel?: string; error?: string }> {
  return postAction('/api/note/new', { name, folder });
}

export function deleteNote(rel: string): Promise<{ ok: boolean; error?: string }> {
  return postAction('/api/note/delete', { rel });
}

export function renameNote(rel: string, name: string): Promise<{ ok: boolean; rel?: string; error?: string }> {
  return postAction('/api/note/rename', { rel, name });
}

export function moveNote(rel: string, folder = ''): Promise<{ ok: boolean; rel?: string; error?: string }> {
  return postAction('/api/note/move', { rel, folder });
}

export function removeNoteFolder(folder: string, destination = ''): Promise<{ ok: boolean; moved?: { from: string; to: string }[]; error?: string }> {
  return postAction('/api/note/folder/remove', { folder, destination });
}

export function renameNoteFolder(folder: string, name: string): Promise<{ ok: boolean; folder?: string; error?: string }> {
  return postAction('/api/note/folder/rename', { folder, name });
}

// store.json annotation, not vault content - see server.py's pin_note().
export function pinNote(rel: string, pinned: boolean): Promise<{ ok: boolean }> {
  return postAction('/api/note/pin', { rel, pinned });
}

File: backend/routes/notes.py
"""Notes routes: read-note GET, and the note CRUD POST dispatch.

Extracted (and re-wired to call through helpers) from make_handler()'s
do_GET/do_POST dispatch chains in the pre-modularization panel/server.py.
"""

import json
from urllib.parse import parse_qs

from backend.collectors.notes import delete_note, move_note, new_note, pin_note, read_note, remove_folder, rename_folder, rename_note, write_note


def handle_get(handler, path, route, snapshot):
    if path == "/api/note":
        rel = (parse_qs(route.query).get("rel") or [""])[0]
        try:
            handler._send(json.dumps(read_note(snapshot.cfg, rel)))
        except Exception as e:
            handler._send(json.dumps({"ok": False, "error": str(e)[:140]}), code=400)
        return True
    return None


def dispatch_post(cfg, path, body):
    if path == "/api/note/save":
        return write_note(cfg, body.get("rel"), body.get("text") or "")
    if path == "/api/note/new":
        return new_note(cfg, body.get("name"), body.get("folder") or "")
    if path == "/api/note/delete":
        return delete_note(cfg, body.get("rel"))
    if path == "/api/note/rename":
        return rename_note(cfg, body.get("rel"), body.get("name"))
    if path == "/api/note/move":
        return move_note(cfg, body.get("rel"), body.get("folder") or "")
    if path == "/api/note/folder/remove":
        return remove_folder(cfg, body.get("folder"), body.get("destination") or "")
    if path == "/api/note/folder/rename":
        return rename_folder(cfg, body.get("folder"), body.get("name"))
    if path == "/api/note/pin":
        return pin_note(cfg, body.get("rel"), bool(body.get("pinned", True)))
    return None
